server/src/main/java/org/opensearch/index/mapper/BooleanFieldMapper.java

```diff
@@ -257,9 +268,76 @@ public DocValueFormat docValueFormat(@Nullable String format, ZoneId timeZone) {
             return DocValueFormat.BOOLEAN;
         }
 
+        @Override
+        public Query termQuery(Object value, QueryShardContext context) {
+            failIfNotIndexedAndNoDocValues();
+            if (!isSearchable()) {
+                return SortedNumericDocValuesField.newSlowExactQuery(name(), Values.TRUE.bytesEquals(indexedValueForSearch(value)) ? 1 : 0);
+            }
+            Query query = new TermQuery(new Term(name(), indexedValueForSearch(value)));
+            if (boost() != 1f) {
+                query = new BoostQuery(query, boost());
+            }
+            return query;
+        }
+
+        @Override
+        public Query termsQuery(List<?> values, QueryShardContext context) {
+            failIfNotIndexedAndNoDocValues();
+            boolean seenTrue = false;
+            boolean seenFalse = false;
+            for (Object value : values) {
+                if (Values.TRUE.equals(indexedValueForSearch(value))) {
+                    seenTrue = true;
+                } else if (Values.FALSE.equals(indexedValueForSearch(value))) {
+                    seenFalse = true;
+                } else {
+                    return new MatchNoDocsQuery("Values did not contain True or False");
+                }
+            }
+            if (seenTrue) {
+                if (seenFalse) {
+                    return new FieldExistsQuery(name());
+                }
+                return termQuery("true", context);
+            }
+            if (seenFalse) {
+                return termQuery("false", context);
+            }
+            return new MatchNoDocsQuery("Values did not contain True or False");
```

Comment:
Whole function can be better written as:

```
@Override
public Query termsQuery(List<?> values, QueryShardContext context) {
            failIfNotIndexedAndNoDocValues();
            int distinct = 0;
            for (Object value : values) {
                if (Values.TRUE.equals(indexedValueForSearch(value))) {
                    distinct |= 2;
                } else if (Values.FALSE.equals(indexedValueForSearch(value))) {
                    distinct |= 1;
                }
                if (distinct == 3) {
                    return new FieldExistsQuery(name());
                }
            }
            switch(distinct) {
                  case 1: return termQuery("false", context);
                  case 2: return termQuery("true", context);
            }

            return new MatchNoDocsQuery("Values did not contain True or False");
}

Reply:
thanks, this makes sense. I updated the logic.